### PR TITLE
feat(gkehub): generate library

### LINKED ISSUE #8010: GKE Hub API: `gkehub.googleapis.com`

### PR DESCRIPTION
**Pull request by scotthart**
fixes #8010

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012)

### CODE REVIEW

File: google/cloud/gkehub/quickstart/README.md [hunk 1]
```diff
@@ -157,7 +157,7 @@ set GRPC_DEFAULT_SSL_ROOTS_FILE_PATH=%cd%\roots.pem
 ```
 
 [bazel-install]: https://docs.bazel.build/versions/main/install.html
-[quickstart-link]: https://cloud.google.com/gkehub/docs/quickstart
+[quickstart-link]: https://cloud.google.com/kubernetes-engine/docs/quickstart
```
Comment (coryan) [inline google/cloud/gkehub/quickstart/README.md:160]: I would use: https://cloud.google.com/anthos/multicluster-management/connect/how-to

File: google/cloud/gkehub/BUILD.bazel [hunk 2]
```diff
@@ -27,12 +27,12 @@ MOCK_HEADER_GLOB = "mocks/*.h"
 cc_library(
     name = "google_cloud_cpp_gkehub",
     srcs = glob(
-        include=[SOURCE_GLOB],
-        exclude=[MOCK_SOURCE_GLOB],
+        include = [SOURCE_GLOB],
```
Comment (coryan) [inline google/cloud/gkehub/BUILD.bazel:30]: Just FYI (and for future reference): seems like you skipped the formatting step after generating the code.

File: google/cloud/gkehub/doc/main.dox [hunk 3]
```diff
@@ -3,7 +3,8 @@
 @mainpage GKE Hub C++ Client Library
 
 An idiomatic C++ client library for the [GKE Hub][cloud-service-docs], a service
-to <UNKNOWN - NO SERVICE CONFIG DOCUMENTATION SUMMARY>
+to handle the registration of many Kubernetes clusters to Google Cloud, and the\
```
Comment (coryan) [inline google/cloud/gkehub/doc/main.dox:6]: nit: backslash at the EOL.

File: ci/etc/expected_install_directories [hunk 4]
```diff
@@ -56,6 +56,10 @@
 ./include/google/cloud/eventarc/publishing
 ./include/google/cloud/eventarc/publishing/v1
 ./include/google/cloud/eventarc/v1
+./include/google/cloud/gkehub
```
Comment (coryan) [inline ci/etc/expected_install_directories:59]: Seems like you are missing a few directories (not in this file, in `ci/.../cmake-install.sh`)

File: google/cloud/gkehub/README.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+* Bazel and CMake builds
+
+## Documentation
+
+* Official documentation about the [GKE Hub][cloud-service-docs] service
+* [Reference doxygen documentation][doxygen-link] for each release of this
+  client library
+* Detailed header comments in our [public `.h`][source-link] files
+
+[cloud-service-docs]: https://cloud.google.com/kubernetes-engine/docs/
```
Comment (coryan) [inline google/cloud/gkehub/README.md:30]: Seems like the right link is https://cloud.google.com/anthos ?

File: google/cloud/gkehub/doc/main.dox [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+For operations that do not return a value the library simply returns
+`google::cloud::Status`.
+
+### Retry, Backoff, and Idempotency Policies.
+
+The library automatically retries requests that fail with transient errors, and
+uses [exponential backoff] to backoff between retries. Application developers
+can override the default policies.
+
+[cloud-service-docs]: https://cloud.google.com/gkehub
```
Comment (coryan) [inline google/cloud/gkehub/doc/main.dox:85]: Broken link, see above for a suggestion.

**Review by coryan — CHANGES_REQUESTED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012)* status: 0 of 40 files reviewed, 5 unresolved discussions (waiting on @coryan)

___
*[ci/etc/expected_install_directories, line 59 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012#-Mtjnj46WM16KS8TOSXf-r1-59:-MtmiExz0JPDzam9NsQt:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/64f5c020a248d11ac627f8b59bd506f1d2814cf2/ci/etc/expected_install_directories#L59)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Seems like you are missing a few directories (not in this file, in `ci/.../cmake-install.sh`)
</blockquote></details>

Done.
___
*[google/cloud/gkehub/BUILD.bazel, line 30 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012#-Mtjnj46WM16KS8TOSXl-r1-30:-MtmiFnpE-R5dbDI04gl:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/64f5c020a248d11ac627f8b59bd506f1d2814cf2/google/cloud/gkehub/BUILD.bazel#L30)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Just FYI (and for future reference): seems like you skipped the formatting step after generating the code.
</blockquote></details>

Done.
___
*[google/cloud/gkehub/README.md, line 30 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012#-Mtjnj46WM16KS8TOSXn-r1-30:-MtmiGPe5BHjjAIiQVd9:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/64f5c020a248d11ac627f8b59bd506f1d2814cf2/google/cloud/gkehub/README.md#L30)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Seems like the right link is https://cloud.google.com/anthos ?
</blockquote></details>

Done.
___
*[google/cloud/gkehub/doc/main.dox, line 85 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012#-Mtjnj46WM16KS8TOSXq-r1-85:-MtmiI-M74BRJqvWbBcz:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/64f5c020a248d11ac627f8b59bd506f1d2814cf2/google/cloud/gkehub/doc/main.dox#L85)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Broken link, see above for a suggestion.
</blockquote></details>

Done.
___
*[google/cloud/gkehub/quickstart/README.md, line 160 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012#-Mtjnj47SOMKc-O57Zpm-r1-160:-MtmiIo7CyR3Fe_AIQro:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/64f5c020a248d11ac627f8b59bd506f1d2814cf2/google/cloud/gkehub/quickstart/README.md#L160)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I would use: https://cloud.google.com/anthos/multicluster-management/connect/how-to 
</blockquote></details>

Done.

**Comment by dbolduc** [comment]
Sorry, I snuck #8013 in ahead of this PR. Can you regenerate the `*_connection.cc` file here and in the `retail` PR? I think `generate-libraries-ci` will break if we don't do this.

**Review by scotthart — COMMENTED** [review]
Rebased

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012#-:-MtmpCh4CjbmqlmUIfMr:b9x2p3v)* status: 0 of 40 files reviewed, 5 unresolved discussions (waiting on @coryan)

File: google/cloud/gkehub/gke_hub_connection.cc [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+            options.get<GkeHubConnectionIdempotencyPolicyOption>()->clone()) {}
+
+  ~GkeHubConnectionImpl() override = default;
+
+  StreamRange<google::cloud::gkehub::v1::Membership> ListMemberships(
+      google::cloud::gkehub::v1::ListMembershipsRequest request) override {
+    request.clear_page_token();
+    auto stub = stub_;
+    auto retry = std::shared_ptr<GkeHubRetryPolicy const>(
+        retry_policy_prototype_->clone());
```
Comment (dbolduc) [inline google/cloud/gkehub/gke_hub_connection.cc:153]: pls regenerate. This should look like `retry_policy()`

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012)* status: 0 of 40 files reviewed, 6 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[google/cloud/gkehub/gke_hub_connection.cc, line 153 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012#-Mtjnj46WM16KS8TOSXt-r3-153:-MtmsdYt6t4V32z_E2vM:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/f8a585eb9aa57473a7a314616cf58dd51b24ca07/google/cloud/gkehub/gke_hub_connection.cc#L153)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

pls regenerate. This should look like `retry_policy()`
</blockquote></details>

Done.

**Comment by dbolduc** [comment]
Thanks. I am happy. I will let @coryan have the final say

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012)* status: 39 of 41 files reviewed, 1 unresolved discussion (waiting on @coryan and @dbolduc)

___
*[ci/etc/expected_install_directories, line 59 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8012#-Mtjnj46WM16KS8TOSXf-r1-59:-MtmwhrA0TMm-5vdzfVQ:b-hpy9w8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/64f5c020a248d11ac627f8b59bd506f1d2814cf2/ci/etc/expected_install_directories#L59)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I am not sure if you added the additional directories by hand or not.  If you did add them manually: the next time somebody runs the `generate-libraries` build your changes here will be overwritten.
</blockquote></details>

Moved additional install directories to cmake-install.sh

**Pull request merged**